rfc/api.md

```diff
@@ -0,0 +1,120 @@
+
+A service can be detached, then its destructor won't block.
+
+All you need to create a service class instance via service manager it its name and prototype. 
+
+```
+auto node = manager.create<T>("node");
+```
+
+The obtained service is initially in disconnected state. To resolve and connect manually you need to call `connect()` method, which returns flagged future. You can always check is the service is in connected state by calling `connected()` method. Note, that this method does passive connection checking, and it can result in false-positive value if the real connection is already broken, but there are no internal callbacks attached for reading or writing at that time.
+
+The service supports auto reconnection on new requests.
+
+To invoke remote method you need to call `invoke<Event>(args...)` method with appropriate event protocol description and its parameters. This one returns you a future of pair of sender/receiver, by manipulating with each of them you can control an upstream and downstream.
+
+```
+template<class T>
+class service {
+    ~service();
+  
+    void detach();
+  
+    auto connected() const noexcept -> bool;
+    auto connect() -> future<void>;
+
+    template<class Event, class... Args>
+    auto invoke(Args&&... args) -> future<std::tuple<sender<T>, receiver<T>>>;
+};
+```
+
+Cocaine upstream and downstream are completely separated entities. In the common case when you send another event in the open channel you make one step forward your protocol transition graph. There are terminate leafs, which means you cannot send another event when you reach it. On the receiver side it is completely identical image - there are receiver state, after reaching which you cannot receive more chunks.
+
+This complex graph structure requires more runtime checks which can be theoretically replaced with static checks, which gives us the following API.
+
+Transmission channel does step forward after invoking `send` method, which accepts event type and its arguments described in protocol `dispatch_type`. It returns the following transmission state, which can be reassigned to the old `tx`, because it's useless now, which gives us an ability to chain requests:
+
+After termination leaf reached you is unable send more chunks, because there won't be such method in terminated sender.
+
+```
+tx = tx.send<C>(args...).get(); // Simple send, can throw if unable to properly pack given arguments or any network error received.
+
+tx.send<C1>(args1...).get().send<C2>(args2...).get().send<C3>(args3...).get(); // Chained.
+```
+
+In the opposite side there is a received, from which you can receive and extract incoming chunks. Unfortunately it should return both the new receiver type and chunk requested.
+
+```
+std::tie(rx, result) = rx.recv().get(); // Can throw on any network error.
```

Comment:
How do you know what type of receiver to return before you actually receive something? Because depending on value stored in `result` the graph may switch to different states, so new receiver may be different too.